components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90

```diff
@@ -3584,334 +3576,608 @@ end subroutine mam_soaexch_1subarea
+
+        logical,  intent(in) :: iscldy ! true if it is a cloudy cell
+
+        real(r8), intent(in) :: qi_vmr(:,:)           ! mass mixing ratios (mmr) [kmol/kmol]
+        real(r8), intent(in) :: qi_del_growth(:,:) !growth in mmr [kmol/kmol]
+
+        real(r8), intent(in), optional :: qcld_vmr(:,:)
+        real(r8), intent(in), optional :: qcld_del_growth(:,:)
+
+        !intent-outs
+        real(r8), intent(out) :: dryvol_a(:), dryvol_c(:)       !dry volumes (before growth) [m3/kmol-air]
+        real(r8), intent(out) :: deldryvol_a(:), deldryvol_c(:) !change in dry volumes [m3/kmol-air]
+
+        integer :: imode
+        integer :: dest_mode
+
+        !For each mode, compute the initial (before growth) dryvolume and the growth in dryvolume
+        do imode = 1, nmode
+           !compute dry volume only for modes participating in inter-modal transfer
+           dest_mode = dest_mode_of_mode(imode)
+           if (dest_mode <= 0) cycle
+
+           !compute dry volumes (before growth) and its change for interstitial aerosols
+           call dryvolume_change(imode, nspec, qi_vmr, qi_del_growth, & !input
+                dryvol_a(imode), deldryvol_a(imode)) !output
+
+           if ( iscldy ) then ! if this grid cell has cloud
+              !if a grid cell is cloudy, clloud borne quantities has to be present
+              if(.not. present(qcld_vmr) .or. .not. present(qcld_del_growth)) then
+                 call endrun('If a grid cell is cloudy, dryvol_c and deldryvol_c should be present: '//errmsg(__FILE__,__LINE__))
+              endif
+              !compute dry volume (before growth) and its change for cloudborne aerosols
+              call dryvolume_change(imode, nspec, qcld_vmr, qcld_del_growth, &!input
+                   dryvol_c(imode), deldryvol_c(imode)) !output
+           end if !iscldy then
+        end do
+
+      end subroutine compute_dryvol_change_in_src_mode
+
+      !----------------------------------------------------------------------
+      !----------------------------------------------------------------------
+
+      subroutine dryvolume_change (imode, nspec, q_vmr, q_del_growth, &!input
+           dryvol, deldryvol) !output
+
+        !intent-ins
+        integer,  intent(in) :: imode           !current mode number
+        integer,  intent(in) :: nspec           !number of species in the current mode
+        real(r8), intent(in) :: q_vmr(:,:)        !volume mixing ratio [kmol/kmol] FIXME: units needs to be reverified
+        real(r8), intent(in) :: q_del_growth(:,:) !change (delta) in volume mixing ratio [kmol/kmol]
+
+        !intent-outs
+        real(r8), intent(out) :: dryvol, deldryvol !dry volume (before growth) and its grwoth [m3/kmol]
+
+        !local variables
+        integer  :: ispec, s_spec_ind, e_spec_ind
+        real(r8) :: tmp_dryvol, tmp_del_dryvol
+
+        !For each mode, we compute a dry volume by combining (accumulating) mass/density for each specie in that mode.
```

Comment:
"Species" is an unusual word that English inherited from Latin. It is both singular and plural ("a species" and "many species" are both valid).